File: src/game/inventory/capacity.ts
import type { PlayerSave } from '../save/types'
import { isGoldCurrencyItem } from './gold'

/** Bag slot cap (each stack or enchanted item uses one slot). */
export const INVENTORY_SLOT_LIMIT = 180

/** Max quantity on a single non-enchanted stack. */
export const INVENTORY_STACK_MAX = Number.MAX_SAFE_INTEGER

export function inventorySlotCount(save: Pick<PlayerSave, 'inventory'>): number {
  return save.inventory.length
}

export function inventorySlotsFree(save: Pick<PlayerSave, 'inventory'>): number {
  return Math.max(0, INVENTORY_SLOT_LIMIT - inventorySlotCount(save))
}

function isUnenchantedMatch(
  stack: PlayerSave['inventory'][number],
  itemId: string,
): boolean {
  return stack.itemId === itemId && !stack.enchantmentId
}

/**
 * Index of the unenchanted pile to grow. Hearted stacks win when both exist.
 * Enchanted items never merge. Returns -1 when a new slot is needed.
 */
export function mergeableStackIndex(
  inventory: PlayerSave['inventory'],
  itemId: string,
  enchantmentId: string | null = null,
): number {
  if (enchantmentId) return -1
  let unfavorited = -1
  for (let i = 0; i < inventory.length; i += 1) {
    const stack = inventory[i]
    if (!stack || !isUnenchantedMatch(stack, itemId)) continue
    if (stack.favorite === true) return i
    if (unfavorited < 0) unfavorited = i
  }
  return unfavorited
}

/** How many of this item can still be added without overflowing slots or stack max. */
export function maxAddableQuantity(
  save: Pick<PlayerSave, 'inventory'>,
  itemId: string,
  enchantmentId: string | null = null,
  _favorite = false,
): number {
  if (isGoldCurrencyItem(itemId) && !enchantmentId) {
    return INVENTORY_STACK_MAX
  }
  if (enchantmentId) {
    return inventorySlotsFree(save)
  }

  const existingIndex = mergeableStackIndex(save.inventory, itemId)
  const existing = existingIndex >= 0 ? save.inventory[existingIndex] : undefined
  const stackRoom = existing
    ? Math.max(0, INVENTORY_STACK_MAX - existing.quantity)
    : inventorySlotsFree(save) > 0
      ? INVENTORY_STACK_MAX
      : 0
  return stackRoom
}

export function canFitItemQuantity(
  save: Pick<PlayerSave, 'inventory'>,
  itemId: string,
  quantity: number,
  enchantmentId: string | null = null,
  favorite = false,
): boolean {
  const want = Math.floor(quantity)
  if (want <= 0) return true
  return maxAddableQuantity(save, itemId, enchantmentId, favorite) >= want
}

File: packages/ik_rules/lib/src/inventory/capacity.dart
import 'dart:math' as math;

import 'package:ik_content/ik_content.dart';

import '../js_compat.dart';
import '../save/generated/save_models.dart';
import 'gold.dart';

/// Bag slot cap; each stack or enchanted item uses one slot.
const int inventorySlotLimit = 180;

/// Max quantity on a single non-enchanted stack, matching `Number.MAX_SAFE_INTEGER`.
const int inventoryStackMax = 9007199254740991;

int inventorySlotCount(PlayerSave save) => save.inventory.length;

int inventorySlotsFree(PlayerSave save) =>
    math.max(0, inventorySlotLimit - inventorySlotCount(save));

/// Index of the unenchanted pile to grow. Hearted stacks win when both exist.
/// Enchanted items never merge. Returns -1 when a new slot is needed.
int mergeableStackIndex(List<InventoryStack> inventory, String itemId, [String? enchantmentId]) {
  if (isNotBlank(enchantmentId)) return -1;
  var unfavorited = -1;
  for (final entry in inventory.indexed) {
    final stack = entry.$2;
    if (stack.itemId != itemId || isNotBlank(stack.enchantmentId)) continue;
    if (stack.favorite == true) return entry.$1;
    if (unfavorited < 0) unfavorited = entry.$1;
  }
  return unfavorited;
}

/// How many of this item can still be added without overflowing slots or the
/// stack maximum.
num maxAddableQuantity(
  PlayerSave save,
  String itemId, [
  String? enchantmentId,
  bool favorite = false,
  GameDatabase? db,
]) {
  if (isGoldCurrencyItem(itemId, db) && isBlank(enchantmentId)) {
    return inventoryStackMax;
  }
  if (isNotBlank(enchantmentId)) {
    return inventorySlotsFree(save);
  }

  final existingIndex = mergeableStackIndex(save.inventory, itemId);
  if (existingIndex >= 0) {
    return math.max(0, inventoryStackMax - save.inventory[existingIndex].quantity);
  }
  return inventorySlotsFree(save) > 0 ? inventoryStackMax : 0;
}

bool canFitItemQuantity(
  PlayerSave save,
  String itemId,
  num quantity, [
  String? enchantmentId,
  bool favorite = false,
  GameDatabase? db,
]) {
  final want = quantity.floor();
  if (want <= 0) return true;
  return maxAddableQuantity(save, itemId, enchantmentId, favorite, db) >= want;
}
